lib/semantics/check-do.cc

```diff
@@ -493,87 +434,137 @@ class DoContext {
     currentStatementSourcePosition_ = doStmt.source;
 
     const parser::Block &block{std::get<parser::Block>(doConstruct.t)};
-    DoConcurrentEnforcement doConcurrentEnforcement{context_};
-    parser::Walk(block, doConcurrentEnforcement);
+    DoConcurrentBodyEnforce doConcurrentBodyEnforce{context_};
+    parser::Walk(block, doConcurrentBodyEnforce);
 
     DoConcurrentLabelEnforce doConcurrentLabelEnforce{context_,
-        doConcurrentEnforcement.labels(), doConcurrentEnforcement.names(),
+        doConcurrentBodyEnforce.labels(), doConcurrentBodyEnforce.names(),
         currentStatementSourcePosition_};
     parser::Walk(block, doConcurrentLabelEnforce);
 
-    auto &loopControl{
+    const auto &loopControl{
         std::get<std::optional<parser::LoopControl>>(doStmt.statement.t)};
-    auto &concurrent{std::get<parser::LoopControl::Concurrent>(loopControl->u)};
-    EnforceConcurrentLoopControl(concurrent, block);
+    const auto &concurrent{
+        std::get<parser::LoopControl::Concurrent>(loopControl->u)};
+    CheckConcurrentLoopControl(concurrent, block);
   }
 
-  void CheckZeroOrOneDefaultNone(
+  using SymbolSet = std::set<const Symbol *>;
+
+  // Return a set of symbols whose names are in a Local locality-spec.  Look
+  // the names up in the scope that encloses the DO construct to avoid getting
+  // the local versions of them.  Then follow the host-, use-, and
+  // construct-associations to get the root symbols
+  SymbolSet GatherLocals(
       const std::list<parser::LocalitySpec> &localitySpecs) const {
-    // C1127
-    int count{0};
-    for (auto &ls : localitySpecs) {
-      if (std::holds_alternative<parser::LocalitySpec::DefaultNone>(ls.u)) {
-        ++count;
-        if (count > 1) {
-          context_.Say(currentStatementSourcePosition_,
-              "only one DEFAULT(NONE) may appear"_en_US);
-          return;
+    SymbolSet symbols;
+    const Scope &scope{
```

Comment:
This is a nitpick, but naming this variable something like `parentScope` would make the intentions of this code slightly more clear. However, you can choose to ignore this suggestion since you already have a comment that clearly tells all that needs to be told.

Reply:
Good point.  I'll make that change.